Adding shader preprocessor support

Chaosus opened the pull request:
A rebased version of https://github.com/godotengine/godot/pull/54690 (by @TheOrangeDay).

Chaosus committed:
Adding shader preprocessor support

Co-authored-by: TheOrangeDay <6472143+TheOrangeDay@users.noreply.github.com>

Ansraer commented:
Thanks for reviving this PR!
iirc exposing defines to the editor inspector was never included in the original PR. Could you take a look at that? Would be really great for plugin authors if users could simply enable/disable shader features in the material editor ui.

Chaosus commented:
> iirc exposing defines to the editor inspector was never included in the original PR. Could you take a look at that?

I will look what I can do...

Chaosus commented:
I've did a cleanup and fixed some bugs (#include is now worked) + added code completion (after `#` symbol and pragma) support. I think its fine to merge for now. Adding advanced features like switching #define in the inspector can be added later.

clayjohn commented:
Wonderful! I'll test out the changes as soon as I can!

CsloudX commented:
SO LOVE this feature！

clayjohn commented:
Great work as always!

I am still not happy with how including works with shader types. For example, If I have a sprite with the following shader:
```
shader_type canvas_item;

#define CLAYD

#include "clay_include.gdshader"

void fragment() {
	COLOR = get_color();
#ifdef CLAY
	COLOR = vec4(0.5, 0.7, 0.95, 1.0);
#endif
}
```
And then the contends of ``clay_include.gdshader`` are:
```
shader_type spatial;

vec4 get_color() {
#ifdef CLAYD
	return vec4(1.0, 0.9, 0.4, 1.0);
#else
	return vec4(1.0);
#endif
}
```
The shader correctly compiles and outputs a color of ``vec4(1.0, 0.9, 0.4, 1.0);``, however, the editor constantly shows an error saying "Constants cannot be modified" pointing to line 8 of the main shader ``COLOR = get_color();``. So the error handling for spatial shaders is being used within a canvas_item shader. 

Alternatively, you have the opposite problem if the spatial shader you include contains a processor function then the built ins will be evaluated as if it were a canvas_item shader.

For example, if the main shader is:
```
shader_type canvas_item;

#include "clay_include.gdshader"

```
and the contents of ``clay_include.gdshader`` are:
```
shader_type spatial;

void fragment() {
	ALBEDO = vec3(1.0);
}
```
When viewing the main shader you get an error ``Unkown identifier in expression "ALBEDO", but when viewing the ``clay_include.gdshader`` shader, there is no error at all.

To me this behaviour is very confusing and quite limiting for users. As pointed out in the previous PR, users will likely want to include the same file in both spatial shaders and canvas_item shaders, so we should support that somehow. 

I can see two options (but you may be able to think of something better):
1. Ban including different shader types, but create a new shader type (called "include") than can be included from anything. I.e. canvas_item can include canvas_item or include, but not spatial, sky, or fog. Fog can include Fog or include etc.
2. Allow including other shader_types. but silently ignore processor functions from other shader types, this way any shader can be included from any other, but it makes it impossible to share processor functions

Chaosus commented:
@clayjohn Will look to fix that on days, rebased for now...

Chaosus commented:
@clayjohn I've added `include` shader mode - made it mandatory for the included files - now they will throw error when you can try to use built-ins from other shader modes inside them, fixed few mode bugs I found. Please check again how it looks.

clayjohn commented:
We discussed this at length on RocketChat and in a subsequent video meeting. 

The consensus was:
1. The general approach here is fine, while it is more complex than a simple preprocessor (which just inserts text from another file) the way this is presented to the user is both simple and effective
2. We should not have a ``shader_type include`` instead, we should create a new resource ``ShaderInclude`` which simply has a single property ``text`` which contains the text of the include shader. ShaderInclude would use a different extension from shaders as well. Perhaps "gdshader.inc". This would not be a full-fledged shader, but rather would be a resource that is editable in the shader editor while editing a proper shader that includes it. ShaderIncludes cannot define processor functions (vertex(), fragment() etc.) doing so will result in an error in the shader that includes them.
3. Shaders can include other shaders from their own type, and they can include ShaderIncludes

Chaosus commented:
@clayjohn I think I'm done with all of these points:

- Removed include shader type
- Added `ShaderInclude` resource type and associate it with *.inc file extension (not sure about `text` property - maybe `code` sounds better? - if not please comment and I will change it back) 
- Shader editor will able to edit them freely, code coloring is correctly working but completion does not
- Added a check to prevent included non-similar shader types

Check again.

clayjohn commented:
@Chaosus You are amazing! I will check the PR right now.

clayjohn commented:
Looks great so far! 

Notes:
1. We should allow users to create new ShaderIncludes in the "File" menu in the Shaders tab
![Screenshot from 2022-06-23 15-18-26](https://user-images.githubusercontent.com/16521339/175424776-69b07e9b-9dc0-4b6d-a506-4735c2e6346e.png)
2. I get a crash after editing a ShaderInclude:
```
================================================================
handle_crash: Program crashed with signal 11
Engine version: Godot Engine v4.0.alpha.custom_build (06ad8fc4807714e3389b19ae015619fd66a906e4)
Dumping the backtrace. Please include this when reporting the bug to the project developer.
[1] /lib/x86_64-linux-gnu/libc.so.6(+0x42520) [0x7ff8eec2f520] (??:0)
[2] CowData<char32_t>::_ref(CowData<char32_t> const&) (??:0)
[3] String::String(String const&) (??:0)
[4] Resource::get_path() const (??:0)
[5] ShaderTextEditor::_validate_script() (??:0)
[6] CodeTextEditor::_text_changed_idle_timeout() (??:0)
[7] void call_with_variant_args_helper<CodeTextEditor>(CodeTextEditor*, void (CodeTextEditor::*)(), Variant const**, Callable::CallError&, IndexSequence<>) (??:0)
[8] void call_with_variant_args<CodeTextEditor>(CodeTextEditor*, void (CodeTextEditor::*)(), Variant const**, int, Callable::CallError&) (??:0)
[9] CallableCustomMethodPointer<CodeTextEditor>::call(Variant const**, int, Variant&, Callable::CallError&) const (??:0)
[10] Callable::call(Variant const**, int, Variant&, Callable::CallError&) const (??:0)
[11] Object::emit_signalp(StringName const&, Variant const**, int) (??:0)
[12] Error Object::emit_signal<>(StringName const&) (??:0)
[13] Timer::_notification(int) (??:0)
[14] Timer::_notificationv(int, bool) (??:0)
[15] Object::notification(int, bool) (??:0)
[16] SceneTree::_notify_group_pause(StringName const&, int) (??:0)
[17] SceneTree::process(double) (??:0)
[18] Main::iteration() (??:0)
[19] OS_LinuxBSD::run() (??:0)
[20] /home/clayjohn/dev/godot/bin/godot.linuxbsd.tools.64.llvm(main+0x1c8) [0x55ea11286df8] (??:0)
[21] /lib/x86_64-linux-gnu/libc.so.6(+0x29d90) [0x7ff8eec16d90] (??:0)
[22] /lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0x80) [0x7ff8eec16e40] (??:0)
[23] /home/clayjohn/dev/godot/bin/godot.linuxbsd.tools.64.llvm(_start+0x25) [0x55ea11286b65] (??:0)
-- END OF BACKTRACE --
================================================================

clayjohn commented on editor/plugins/shader_editor_plugin.cpp:
```diff
@@ -561,41 +826,132 @@ void ShaderEditor::edit(const Ref<Shader> &p_shader) {
+	}
+
+	shader_inc = p_shader_inc;
+	shader = nullptr;
```
I manually changed this to ``shader= Ref<Shader>();`` Im not sure if it is correct to do so.

Chaosus commented on editor/plugins/shader_editor_plugin.cpp:
```diff
@@ -561,41 +826,132 @@ void ShaderEditor::edit(const Ref<Shader> &p_shader) {
 	}
 
 	shader = p_shader;
+	shader_inc = nullptr;
```
Ah, I've already did it at the branch, you've downloaded previous commit.

Chaosus replied:
Done

Chaosus commented:
@clayjohn Check again - I've added menu button for create include file and possibly fixed a crash you've mentioned (I'm not sure about it since you didn't posted exact line of where in `_validate_script` it happen). I've also added a check to restrict user to define preprocessor directives before `shader_type` (I found it error-prone and leads to hard-detected errors)

Chaosus commented:
@clayjohn Ok, I think I've fixed these issues - try again.

Paulb23 commented on editor/code_editor.cpp:
```diff
@@ -1593,6 +1593,10 @@ void CodeTextEditor::set_error_pos(int p_line, int p_column) {
 	error_column = p_column;
 }
 
+void CodeTextEditor::get_error_pos(Point2i &r_pos) const {
+	r_pos = Point2i(error_line, error_column);
+}
```
```suggestion
Point2i CodeTextEditor::get_error_pos() const {
	return Point2i(error_line, error_column);
}
```

Chaosus replied:
Done

Zylann commented:
I think `*.inc` is not specific enough for shader includes, `*.gdshaderinc` would be better IMO, even though it appears long

Chaosus commented:
> I think *.inc is not specific enough for shader includes, *.gdshaderinc would be better IMO, even though it appears long.

If @reduz of @clayjohn agreed with it - I will change it

> How do we set the defines? Do they automatically appear as toggles in ShaderMaterial for all the #if X found in the final shader? Or maybe do we "export" them explicitly with some special syntax?

There are not adding toggle to the inspector, for now (this feeature is already requested at the beginning of this PR topic). This PR bring a huge mess of code already, I would add it later.

clayjohn commented:
> > I think *.inc is not specific enough for shader includes, *.gdshaderinc would be better IMO, even though it appears long.
> 
> If @reduz or @clayjohn agreed with it - I will change it

I think we should change to ".gdshaderinc". ".gdshader.inc" was reduz' original suggestion after all.

Chaosus commented:
> I think we should change to ".gdshaderinc". ".gdshader.inc" was reduz' original suggestion after all.

Done

clayjohn commented:
Amazing work! I tested out before the name change and everything appeared to be working well. I will do an in-depth code review tomorrow, but I am fairly sure this will be ready to go :)

reduz commented on servers/rendering/shader_compiler.cpp:
```diff
@@ -1320,12 +1321,39 @@ Error ShaderCompiler::compile(RS::ShaderMode p_mode, const String &p_code, Ident
 	info.shader_types = ShaderTypes::get_singleton()->get_types();
 	info.global_variable_type_func = _get_variable_type;
 
+	// Cheat the context here for now.
```
This code __really__ should not be here, the compiler should give you the right file and line when there is an error, and it should not depend on the preprocessor at this point.

reduz commented on servers/rendering/shader_preprocessor.cpp:
```diff
@@ -0,0 +1,1317 @@
+		return;
+	}
+
+	Ref<Shader> shader = Object::cast_to<Shader>(*res);
```
As we discussed, I think including shaders should be avoided, and only ShaderInclude files should be allowed.

reduz commented on servers/rendering/shader_preprocessor.cpp:
```diff
@@ -0,0 +1,1317 @@
+
+	String result;
+	processor.preprocess(state, result);
+	add_to_output(result);
```
As I suggested, I think it would be much simpler that its the ShaderLanguage tokenizer that gets rid of correcting file and line by inserting special tokens for this. This way you can entirely discard the preprocessor after it did its job.

Chaosus replied:
Not clearly understand what you mean here and how it can be done. Can you post a sample code?

reduz commented on servers/rendering/shader_preprocessor.cpp:
```diff
@@ -0,0 +1,1317 @@
+		}
+	}
+
+	if (is_shader) {
```
I have to say this code ended up being pretty simple in the end, so maybe you could still leave it be for include files (allowing to optionally specify shader_type if you want to edit the thing standalone?)

reduz commented on editor/inspector_dock.h:
```diff
@@ -141,6 +143,7 @@ class InspectorDock : public VBoxContainer {
 	void go_back();
 	void edit_resource(const Ref<Resource> &p_resource);
 	void open_resource(const String &p_type);
+	void open_resource(const Vector<String> &p_types);
```
I dont quite understand what this is for yet

Chaosus replied:
I did it to allow multiple type of resource to be loaded by "Load Shader" menu option. Because `Shader` and `ShaderInclude` are different kind of resources and I wanted to load either of them using `FileDialog`.

reduz commented:
My feeling after seeing this PR again and again is that its too complicated in many aspects where it should really not be. My suggestion for simplifying and improving it would be as follows:

### 1: Remove the ability to edit the includes as sub-editors, allow .gshaderinc to be opened as any other shader
![image](https://user-images.githubusercontent.com/6265307/175805247-24230e6b-71e5-4cb2-a692-75ca515be9a8.png)
I know this looks cool but let's be honest. This was made for the previous shader editor where you could only edit one shader at a time. This is no longer the case now, so this feature is very redundant. Removing it simplifies the shader editor significantly.
Additionally, as I saw in code, the ability to optionally put a shader_type in an include to get proper compilation errors was so simple for the preprocessor to remove, that should let you edit the includes in a separate item in the list with proper code completion and without any issues.
### 2: Make the preprocessor really a preprocessor.
I know I insist a lot in this, but There is a lot of quite ugly code due to this. As a rule:
1. The preprocessor should be used **ONLY** from  ShaderLanguage and internally (unexposed). No other files should reference or use it (which is the case of ShaderCompiler now and ShaderEditorPlugin.
2. The preprocessor include files should be passed to the tokenizer as special tokens, which should maintain an include stack and use this for proper error reporting. I can help implementing this if you want, it can be done last as cleanup when everything else is working.
### 3: Settle on single extension (.gshaderinc or something else)
Godot is not designed to handle multiple extensions (.gshader.inc will not work properly), and changing it only to support this would be an immense amount of work that is unjustified.

reduz commented on editor/inspector_dock.cpp:
```diff
@@ -211,15 +211,10 @@ void InspectorDock::_new_resource() {
-	List<String> extensions;
-	ResourceLoader::get_recognized_extensions_for_type(p_type, &extensions);
-
+void InspectorDock::_load_resources_ext(const List<String> &p_extensions) {
```
I dont quite understand what this is for. Can we do without it?

Chaosus replied:
I did it to allow Loading multiple file extensions by using `Load Shader` option. Since Shader and ShaderInclude are different kind of resources they need to be supported by the file dialog.

reduz replied:
I see, maybe its probably better to do a custom dialog for this then

Chaosus replied:
I can revert this change if you don't want it and add a separate menu option... Creating a whole new dialog just for this... I don't know.. - maybe in the other PR?

reduz commented on editor/scene_tree_dock.h:
```diff
@@ -296,8 +298,12 @@ class SceneTreeDock : public VBoxContainer {
 	void perform_node_renames(Node *p_base, HashMap<Node *, NodePath> *p_renames, HashMap<Ref<Animation>, HashSet<int>> *r_rem_anims = nullptr);
-	SceneTreeEditor *get_tree_editor() { return scene_tree; }
-	EditorData *get_editor_data() { return editor_data; }
+	SceneTreeEditor *get_tree_editor() {
```
Looks like clang format did something to this file

Calinou commented:
> `.gshaderinc`

I'd suggest `.gdshaderinc`, not `.gshaderinc`, as standard shader files use `.gdshader`.

Chaosus commented:
> I'd suggest .gdshaderinc, not .gshaderinc, as standard shader files use .gdshader.

I've already did `.gdshaderinc` as you've suggested
